File: sdk/python/arguing_client.py
"""
Arguing API Python Client SDK

A simple Python client for the Arguing REST API.
Uses only Python standard library - no external dependencies required.

Example usage:
    client = ArguingClient(base_url='http://localhost:3000', api_key='your-api-key')
    graph = client.get_graph()
    nodes = client.get_nodes()
    node = client.get_node('1')
"""

import json
import urllib.request
import urllib.error
from typing import Any, Optional


class ArguingAPIError(Exception):
    """Exception raised for API errors."""

    def __init__(self, status_code: int, message: str):
        self.status_code = status_code
        self.message = message
        super().__init__(f"API Error {status_code}: {message}")


class ArguingClient:
    """Client for the Arguing REST API."""

    def __init__(self, base_url: str = 'http://localhost:3000', api_key: str = ''):
        self.base_url = base_url.rstrip('/')
        self.api_key = api_key
        self.api_base = f"{self.base_url}/api/v1"

    def _request(self, path: str, method: str = 'GET', body: Optional[dict] = None,
                 accept: str = 'application/json') -> Any:
        url = f"{self.api_base}{path}"
        headers = {
            'Authorization': f'Bearer {self.api_key}',
            'Accept': accept,
            'Content-Type': 'application/json',
        }
        data = json.dumps(body).encode('utf-8') if body else None
        req = urllib.request.Request(url, data=data, headers=headers, method=method)
        try:
            with urllib.request.urlopen(req) as response:
                raw = response.read().decode('utf-8')
                if 'json' in response.headers.get('Content-Type', ''):
                    return json.loads(raw)
                return raw
        except urllib.error.HTTPError as e:
            error_body = e.read().decode('utf-8')
            try:
                error_data = json.loads(error_body)
                msg = error_data.get('message', error_body)
            except Exception:
                msg = error_body
            raise ArguingAPIError(e.code, msg) from e

    def get_nodes(self, accept: str = 'application/json') -> list:
        """Get all nodes in the argument graph."""
        result = self._request('/nodes', accept=accept)
        return result.get('nodes', []) if isinstance(result, dict) else result

    def get_node(self, node_id: str) -> dict:
        """Get a specific node by ID."""
        return self._request(f'/nodes/{node_id}')

    def get_edges(self) -> list:
        """Get all edges in the argument graph."""
        result = self._request('/edges')
        return result.get('edges', [])

    def get_edges_by_source(self, source_id: str) -> list:
        """Get all edges from a specific source node."""
        result = self._request(f'/edges/filter?source={source_id}')
        return result.get('edges', [])

    def get_edges_by_target(self, target_id: str) -> list:
        """Get all edges to a specific target node."""
        result = self._request(f'/edges/filter?target={target_id}')
        return result.get('edges', [])

    def get_graph(self) -> dict:
        """Get the full argument graph (nodes and edges)."""
        return self._request('/graph')

    def register_webhook(self, url: str, events: list, secret: Optional[str] = None) -> dict:
        """Register a webhook for event notifications."""
        body = {'url': url, 'events': events}
        if secret:
            body['secret'] = secret
        return self._request('/webhooks', method='POST', body=body)

    def list_webhooks(self) -> list:
        """List all registered webhooks."""
        result = self._request('/webhooks')
        return result.get('webhooks', [])

    def delete_webhook(self, webhook_id: str) -> None:
        """Delete a registered webhook."""
        self._request(f'/webhooks/{webhook_id}', method='DELETE')

    def graphql_query(self, query: str, variables: Optional[dict] = None) -> dict:
        """Execute a GraphQL query."""
        body = json.dumps({'query': query, 'variables': variables or {}}).encode('utf-8')
        headers = {
            'Authorization': f'Bearer {self.api_key}',
            'Content-Type': 'application/json',
            'Accept': 'application/json',
        }
        url = f"{self.base_url}/graphql"
        request = urllib.request.Request(url, data=body, headers=headers, method='POST')
        with urllib.request.urlopen(request) as response:
            return json.loads(response.read().decode('utf-8'))

File: sdk/javascript/arguing_client.js
/**
 * Arguing API JavaScript/Node.js Client SDK
 *
 * A simple client for the Arguing REST API.
 *
 * Supports both Node.js (using built-in fetch in Node 18+) and browser environments.
 *
 * @example
 * const client = new ArguingClient({ baseUrl: 'http://localhost:3000', apiKey: 'your-api-key' });
 * const graph = await client.getGraph();
 * const nodes = await client.getNodes();
 */

'use strict';

class ArguingAPIError extends Error {
    constructor(statusCode, message) {
        super(`API Error ${statusCode}: ${message}`);
        this.name = 'ArguingAPIError';
        this.statusCode = statusCode;
    }
}

class ArguingClient {
    /**
     * @param {object} options
     * @param {string} [options.baseUrl='http://localhost:3000'] - Base URL of the API server
     * @param {string} [options.apiKey=''] - API key for authentication
     */
    constructor({ baseUrl = 'http://localhost:3000', apiKey = '' } = {}) {
        this.baseUrl = baseUrl.replace(/\/$/, '');
        this.apiKey = apiKey;
        this.apiBase = `${this.baseUrl}/api/v1`;
    }

    async _request(path, { method = 'GET', body, accept = 'application/json' } = {}) {
        const url = `${this.apiBase}${path}`;
        const headers = {
            'Authorization': `Bearer ${this.apiKey}`,
            'Accept': accept,
            'Content-Type': 'application/json',
        };

        const options = { method, headers };
        if (body !== undefined) {
            options.body = JSON.stringify(body);
        }

        const response = await fetch(url, options);

        if (response.status === 204) return null;

        const text = await response.text();
        let data;
        try {
            data = JSON.parse(text);
        } catch {
            data = text;
        }

        if (!response.ok) {
            const message = (data && data.message) ? data.message : String(data);
            throw new ArguingAPIError(response.status, message);
        }

        return data;
    }

    /** Get all nodes in the argument graph */
    async getNodes(accept = 'application/json') {
        const result = await this._request('/nodes', { accept });
        return result?.nodes ?? result;
    }

    /** Get a specific node by ID */
    async getNode(id) {
        return this._request(`/nodes/${encodeURIComponent(id)}`);
    }

    /** Get all edges in the argument graph */
    async getEdges() {
        const result = await this._request('/edges');
        return result?.edges ?? result;
    }

    /** Get edges filtered by source node ID */
    async getEdgesBySource(sourceId) {
        const result = await this._request(`/edges/filter?source=${encodeURIComponent(sourceId)}`);
        return result?.edges ?? result;
    }

    /** Get edges filtered by target node ID */
    async getEdgesByTarget(targetId) {
        const result = await this._request(`/edges/filter?target=${encodeURIComponent(targetId)}`);
        return result?.edges ?? result;
    }

    /** Get the full argument graph (nodes and edges) */
    async getGraph(accept = 'application/json') {
        return this._request('/graph', { accept });
    }

    /** Register a webhook for event notifications */
    async registerWebhook(url, events, secret) {
        const body = { url, events };
        if (secret) body.secret = secret;
        return this._request('/webhooks', { method: 'POST', body });
    }

    /** List all registered webhooks */
    async listWebhooks() {
        const result = await this._request('/webhooks');
        return result?.webhooks ?? result;
    }

    /** Delete a registered webhook by ID */
    async deleteWebhook(webhookId) {
        return this._request(`/webhooks/${encodeURIComponent(webhookId)}`, { method: 'DELETE' });
    }

    /** Execute a GraphQL query */
    async graphqlQuery(query, variables = {}) {
        const url = `${this.baseUrl}/graphql`;
        const response = await fetch(url, {
            method: 'POST',
            headers: {
                'Authorization': `Bearer ${this.apiKey}`,
                'Content-Type': 'application/json',
                'Accept': 'application/json',
            },
            body: JSON.stringify({ query, variables }),
        });
        const data = await response.json();
        if (data.errors) {
            throw new ArguingAPIError(400, data.errors.map(e => e.message).join(', '));
        }
        return data;
    }
}

// CommonJS export for Node.js
if (typeof module !== 'undefined' && module.exports) {
    module.exports = { ArguingClient, ArguingAPIError };
}
